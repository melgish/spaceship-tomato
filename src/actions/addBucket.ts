'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
