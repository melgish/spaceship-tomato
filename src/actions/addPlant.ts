'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
