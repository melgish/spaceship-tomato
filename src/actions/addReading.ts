'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
