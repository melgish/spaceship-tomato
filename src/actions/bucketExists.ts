'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
