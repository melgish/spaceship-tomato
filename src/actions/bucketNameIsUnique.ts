'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
