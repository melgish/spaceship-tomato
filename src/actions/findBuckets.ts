'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
