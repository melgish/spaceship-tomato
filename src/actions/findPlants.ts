'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of plants from the database.
 * @returns An array of plants.
 */
export async function findPlants() {
  return await prisma.plant.findMany({
    orderBy: [{ name: 'asc' }],
  })
}

export type Plants = Awaited<ReturnType<typeof findPlants>>
export type Plant = Plants[number]
