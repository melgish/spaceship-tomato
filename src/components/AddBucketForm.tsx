'use client'
import { useId } from 'react'
import { useForm } from 'react-hook-form'
import { zodResolver } from '@hookform/resolvers/zod'

import ErrorMessage from '@/components/ErrorMessage'
import ResetButton from '@/components/ResetButton'
import SubmitButton from '@/components/SubmitButton'
import { addBucket } from '@/actions/addBucket'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'
import type { IdAndName } from '@/lib/idAndName'
import { notify } from '@/lib/notify'

const defaultValues: AddBucket = { name: '', plantId: 0 }
const resolver = zodResolver(addBucketSchema)

type Props = Readonly<{ plants: IdAndName[] }>

export default function AddBucketForm({ plants }: Props) {
  const id = useId()

  const clear = () => {
    reset(defaultValues)
  }

  const save = async (data: AddBucket) => {
    if (notify(await addBucket(data))) {
      clear()
    }
  }

  const { handleSubmit, formState, register, reset } = useForm<AddBucket>({
    defaultValues,
    resolver,
    shouldFocusError: true,
  })

  return (
    <form className="flex gap-x-3" noValidate onSubmit={handleSubmit(save)}>
      <div className="required field">
        <label htmlFor={`${id}-name`}>Name</label>
        <input
          {...register('name')}
          autoComplete="off"
          id={`${id}-name`}
          maxLength={24}
          placeholder="Ex. #1"
          required
          size={24}
          type="text"
        />
        <ErrorMessage error={formState.errors?.name} />
      </div>
      <div className="field">
        <label htmlFor={`${id}-plantId`}>Plant</label>
        <select
          {...register('plantId')}
          className="py-1.5"
          style={{ width: '32ch' }}
          id={`${id}-plantId`}
        >
          <option value={0}>None</option>
          {plants.map((p: IdAndName) => (
            <option key={p.id} value={p.id}>
              {p.name}
            </option>
          ))}
        </select>
        <ErrorMessage error={formState.errors.plantId} />
      </div>
      <div className="flex items-center gap-x-2">
        <SubmitButton disabled={formState.isSubmitting} />
        <ResetButton disabled={formState.isSubmitting} onClick={clear} />
      </div>
    </form>
  )
}
