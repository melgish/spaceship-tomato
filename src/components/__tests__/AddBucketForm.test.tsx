import { Mock, beforeEach, describe, expect, it, vi } from 'vitest'
import { act, fireEvent, render, screen } from '@testing-library/react'
import AddBucketForm from '../AddBucketForm'
import { addBucket } from '@/actions/addBucket'
import { bucketNameIsUnique } from '@/actions/bucketNameIsUnique'

const plants = [
  { id: 1, name: 'Test Plant 1', maxPh: 6.5, minPh: 7.1 },
  { id: 2, name: 'Test Plant 2', maxPh: 6.8, minPh: 7.2 },
]

vi.mock('@/actions/bucketNameIsUnique')
vi.mock('@/actions/addBucket')
vi.mock('react-hot-toast')

beforeEach(() => {
  // Mock deep validation
  ;(bucketNameIsUnique as Mock).mockResolvedValue(true)
  ;(addBucket as Mock).mockResolvedValue({ ok: true, message: 'Good' })
})

it('should render', () => {
  const { container } = render(<AddBucketForm plants={plants} />)

  expect(container).toMatchSnapshot()
})

describe('when invalid and submit is clicked', () => {
  it('should report validation errors', async () => {
    const { container } = render(<AddBucketForm plants={plants} />)

    await act(() => fireEvent.click(screen.getByRole('button', { name: 'Submit' })))

    expect(container).toMatchSnapshot()
    expect(addBucket).not.toBeCalled()
  })
})

describe('when valid and submit is clicked', () => {
  it('should submit data', async () => {
    const { container } = render(<AddBucketForm plants={plants} />)

    const name = screen.getByLabelText(/name/i)
    const change = { target: { value: 'My Test Bucket' } }
    await act(() => fireEvent.change(name, change))

    const plant = screen.getByLabelText(/plant/i)
    const select = { target: { value: 0 } }
    await act(() => fireEvent.change(plant, select))

    const submit = screen.getByRole('button', { name: /submit/i })
    await act(() => fireEvent.click(submit))

    expect(container).toMatchSnapshot()
    expect(addBucket).toBeCalledWith({ name: 'My Test Bucket', plantId: 0 })
  })
})
