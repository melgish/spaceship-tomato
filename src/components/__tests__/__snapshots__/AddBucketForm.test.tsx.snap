// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r0:-name"
      >
        Name
      </label>
      <input
        autocomplete="off"
        id=":r0:-name"
        maxlength="24"
        name="name"
        placeholder="Ex. #1"
        required=""
        size="24"
        type="text"
      />
    </div>
    <div
      class="field"
    >
      <label
        for=":r0:-plantId"
      >
        Plant
      </label>
      <select
        class="py-1.5"
        id=":r0:-plantId"
        name="plantId"
        style="width: 32ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Test Plant 1
        </option>
        <option
          value="2"
        >
          Test Plant 2
        </option>
      </select>
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when invalid and submit is clicked > should report validation errors 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r1:-name"
      >
        Name
      </label>
      <input
        autocomplete="off"
        id=":r1:-name"
        maxlength="24"
        name="name"
        placeholder="Ex. #1"
        required=""
        size="24"
        type="text"
      />
      <div
        class="error text-rose-800"
      >
        Name is required
      </div>
    </div>
    <div
      class="field"
    >
      <label
        for=":r1:-plantId"
      >
        Plant
      </label>
      <select
        class="py-1.5"
        id=":r1:-plantId"
        name="plantId"
        style="width: 32ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Test Plant 1
        </option>
        <option
          value="2"
        >
          Test Plant 2
        </option>
      </select>
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;

exports[`when valid and submit is clicked > should submit data 1`] = `
<div>
  <form
    class="flex gap-x-3"
    novalidate=""
  >
    <div
      class="required field"
    >
      <label
        for=":r2:-name"
      >
        Name
      </label>
      <input
        autocomplete="off"
        id=":r2:-name"
        maxlength="24"
        name="name"
        placeholder="Ex. #1"
        required=""
        size="24"
        type="text"
      />
    </div>
    <div
      class="field"
    >
      <label
        for=":r2:-plantId"
      >
        Plant
      </label>
      <select
        class="py-1.5"
        id=":r2:-plantId"
        name="plantId"
        style="width: 32ch;"
      >
        <option
          value="0"
        >
          None
        </option>
        <option
          value="1"
        >
          Test Plant 1
        </option>
        <option
          value="2"
        >
          Test Plant 2
        </option>
      </select>
    </div>
    <div
      class="flex items-center gap-x-2"
    >
      <button
        class="emerald"
        type="submit"
      >
        Submit
      </button>
      <button
        class="slate"
        type="button"
      >
        Clear
      </button>
    </div>
  </form>
</div>
`;
